Enable sorting for generative tests

benbc opened the pull request:
The driver for this is to allow us to do more thorough testing (fixes #461). Previously we couldn't test sorting with the generative tests because in some cases the sort order differed between databases (and even between runs for a given database).

To enable that testing this PR adds a small and possibly user-invisible feature that makes the results of all sorts stable even in cases where the sort is under-specified by the dataset definition.

There are two things that I'm not sure about in this implementation and would welcome feedback on.

1. Should the new transformation be separate or should it be rolled into the existing one. My weak preference is to separate them because they exist for different reasons, but their implementation ties them together so there is an argument for treating them as a single transformation.

2. The unit tests for the transformations are pretty unreadable (and even hard to write). Should we introduce some kind of QM builder tooling to make tests like this more readable?

evansd reviewed (approved):
It's very pleasing that this doesn't even touch the query engines.

I've added one, annoying, comment on the boolean handling stuff but I haven't said much beyond that. There were a few comments which I had to go back and delete because you addressed them in later commits.

I wonder if we might end up, further down the road, handling all the sort lifting here rather than in the query engine, which I think might simplify both this code _and_ the query engine. But I don't think it's worth thinking about now.

> Should the new transformation be separate or should it be rolled into the existing one. My weak preference is to separate them because they exist for different reasons, but their implementation ties them together so there is an argument for treating them as a single transformation.

My preferences here are reasonably weak as well. For what it's worth, I tend to swing the other way because I feel like the sort handling code as a whole would be easier to understand if was handled in one go. I think anyone working on any of the sort code would need to understand and be mindful of both parts, and so I don't see that separating them aids comprehension. But I've no real objection to the code as it is; and I'm happy to go with your preferences here given that you've had to write it.

> The unit tests for the transformations are pretty unreadable (and even hard to write). Should we introduce some kind of QM builder tooling to make tests like this more readable?

It's the inside-out ordering that makes them so hard to take in, isn't it? Did you see that [attempt](https://gist.github.com/evansd/924c7c5fe9f9476215df453482a29dd4) I made at a fluent query model? Leaving aside the magical implementation, the basic idea of being able to write `foo.operation(bar)` rather than `Operation(foo, bar)` seems to me to go a long way to increasing readability while keeping very close to "raw" query model. We could implement something like this the tedious, magic-free way if we thought it was worthwhile?

inglesp commented on databuilder/query_model.py:
```diff
@@ -49,7 +50,7 @@
 T = TypeVar("T")
 Numeric = TypeVar("Numeric", int, float)
-Comparable = TypeVar("Comparable", int, float, str, date)
+Comparable = TypeVar("Comparable", int, float, str, date, BaseCode)
```
Commenting here to respond to the commit message:

> This isn't something that researchers will ever want to do, but it
> allows us to use code columns for tie-breaking sorts to ensure sort
> stability.

I think users may want to do this if they want to ensure exactly the same output between runs.  I think we ensure sort stability in parts of cohort-extractor for this reason.

benbc replied:
My claim is that researchers don't every really want to sort by code for its own sake. They may want to have sort stability, but this isn't the commit that gives that.

inglesp replied:
Makes sense.

inglesp commented on databuilder/query_model_transforms.py:
```diff
@@ -31,72 +36,49 @@ def apply_transforms(variables):
-    # We're only expecting nodes which reference a single input
-    assert get_input_nodes(target) == [new_ref]
+
+def get_all_nodes(variables):
```
I'm unkeen on having functions called `all_nodes` and `get_all_nodes` in the same module, without a docstring indicating why they're different and what they're for.

benbc replied:
:+1: thanks

inglesp commented on databuilder/query_model_transforms.py:
```diff
@@ -31,72 +36,49 @@ def apply_transforms(variables):
+
+def get_dependers(node, nodes):
+    dependers = []
+    for other_node in nodes:
```
Given that this is called from inside a loop over `nodes`, this has quadratic behaviour.  Maybe ok for now?

benbc replied:
It does. I'm v happy to replace with a single pass and a dict if you prefer.

We have a _lot_ of QM graph traversal in the codebase. Could be worth doing some optimization at some point, but I'm not sure this is particularly like to be the worst offender.

inglesp commented on databuilder/query_model_transforms.py:
```diff
@@ -31,72 +36,49 @@ def apply_transforms(variables):
+    return nodes
+
+
+def get_dependers(node, nodes):
```
Docstring?

benbc replied:
:+1:

inglesp commented on databuilder/query_model_transforms.py:
```diff
@@ -68,10 +69,34 @@ def add_selected_columns_to_pick_row(nodes):
+    doesn't help our users but makes testing a bit easier.
+    """
+    for node in nodes:
+        if not isinstance(node, PickOneRowPerPatientWithColumns):
```
Why would this ever be hit?

benbc replied:
Each transformation inspects every single node.

inglesp commented on tests/unit/test_query_model_transforms.py:
```diff
@@ -59,3 +59,79 @@ def test_pick_one_row_per_patient_transform():
+    by_i1 = Sort(events, SelectColumn(events, "i1"))
+    variables = dict(
+        v=SelectColumn(
+            PickOneRowPerPatient(source=by_i1, position=Position.FIRST), "i2"
```
Might help with legibility here and elsewhere if you put a trailing comma in this line, and then black would reformat it so that `"i2"` was on a new line.  This would make it clearer (to me) that `"i2"` is the second argument of the `SelectColumn`.

benbc replied:
Yeah, I think that's a bit better. I think it's a fools errand to try to make these beautiful without more tooling though.

inglesp commented on tests/unit/test_query_model_transforms.py:
```diff
@@ -59,3 +59,79 @@ def test_pick_one_row_per_patient_transform():
+        ),
+    )
+
+    assert transformed == expected_variables
```
Silly thought: could we write these tests in such a way that we could easily extract the input and expected output, and then as a post-processing step produce ascii diagrams of the two graphs, to make it easier to see what the transformation should do?

benbc replied:
That does sound helpful, but I'm skeptical that the effort would pay off at least until we have more transformations.

benbc commented:
@evansd I'll merge the transforms. I think you're weak preferences should hold sway over mine in this codebase.

benbc commented:
@evansd My inclination with the QM writing code is to leave it to a possible future piece of standalone work that the team can prioritize. Are you happy with that?

benbc commented:
@evansd I forgot something! I put the sortable-codes commit in just to get it working, but it wasn't my intention to force it through without more discussion. Do you want to suggest an alternative?

benbc commented on databuilder/query_model_transforms.py:
```diff
@@ -131,6 +134,13 @@ def get_immediate_sorts(node):
+def make_sortable(col):
+    if get_series_type(col) == bool:
+        # Some databases can't sort booleans (including SQL Server), so we map them to integers
+        return Case(cases={col: Value(1)}, default=Value(0))
```
Thank you. I did think about nulls at some point, but then forgot about it.

inglesp commented on tests/spec/sort_and_pick/test_sort_extends_to_all_columns_when_underspecified.py:
```diff
@@ -0,0 +1,24 @@
+    spec_test(
+        table_data,
+        e.sort_by(e.i1, e.i2).first_for_patient().i3,
+        {1: 102, 2: 500},
```
I can't see where the new behaviour is actually tested here.  The sort order is entirely specified by the values in `i1` and `i2`, isn't it?

benbc replied:
Yes, you're quite right. This test underwent last-minute revision that has not been kind to it. Thank you.

inglesp reviewed (approved):
Thanks for doing this.  A few suggestions, but all looks sensible.

This is at the limit of what I can keep in my head a the moment though.  As you've identified, we should think about how we can make the tests more readable.

benbc commented:
@evansd 

> I wonder if we might end up, further down the road, handling all the sort lifting here rather than in the query engine, which I think might simplify both this code and the query engine.

Yes, I had the same thought.

benbc commented on databuilder/query_engines/in_memory.py:
```diff
@@ -100,7 +103,7 @@ def visit_Sort(self, node):
         return source.sort(sort_index)
 
-    def visit_PickOneRowPerPatient(self, node):
+    def visit_PickOneRowPerPatientWithColumns(self, node):
```
No. This latest change is not a transformation, it changes the behaviour by stabilizing under-specified sorts.

inglesp replied:
Ah yes, of course.

evansd commented:
@benbc

> My inclination with the QM writing code is to leave it to a possible future piece of standalone work that the team can prioritize. Are you happy with that?

Yes, I agree.

---

> I forgot something! I put the sortable-codes commit in just to get it working, but it wasn't my intention to force it through without more discussion. Do you want to suggest an alternative?

I think there are two questions here: the important one is whether we want to allow users to sort on code columns, as that becomes part of the public API we need to support — whether we document it or not. There's a secondary question about the implementation (what should import from what) which is something we can revisit later.

I'm coming tentatively to the view that we should allow sorting on **all columns**, including both codes and booleans.

If a dataset definition has an under-specified sort then its results are arbitrary. This PR helpfully ensures that it's _consistently_ arbitrary, which is an important feature; but I think ideally any tie-breaks required should be specified in the dataset definition itself so they can be reviewed for possible statistical implications.

We _could_ require every sort to be fully specified, but that feels quite user-unfriendly. (Maybe we could make it a warning or something though?) But I think we should at least provide the tools necessary to fully specify the sort. And that means being able to sort on all columns.

That's a long-winded way to say: I'm happy with this PR as it is on this point, and I'm thinking we should go further.

evansd reviewed (approved):
Remember when we thought this might be a one-line change? :grinning: 

I'm generally happy with this (and of course delighted with the behaviour it adds). The thing which gives me most pause is the change to the query engine to require filters before sorts. It seems to me to not simplify the code all that much. And the `pop_filters`/`pop_sorts` functions will give silently incorrect results if the implicit invariant they rely on is violated.

The mutating-the-immutable approach is also starting to feel not quite as quick a hack as it started out as. But then again, the various IdentityCollections are conceptually very straightforward even if they require some boilerplate. And once we hide them away in a utils module the transform code itself is pretty simple.

inglesp commented on databuilder/query_model_transforms.py:
```diff
@@ -28,75 +42,184 @@ class PickOneRowPerPatientWithColumns(PickOneRowPerPatient):
-        # Update any references to the old node to point to the new one
-        for ref_node in references:
-            update_references(ref_node, node, new_node)
+    return col
```
Do we need a test to demonstrate the new behaviour, that boolean values are sorted as we expect?

benbc replied:
We do! Added, thanks.

inglesp reviewed (commented):
I'm afraid I don't have the time to do a thorough review here.

benbc commented:
@evansd I'm all ready to rip out the change here to canonicalize the sort/filter order in the transforms. But I wonder if I've missed some opportunities to simplify the query engine based on that canonicalization which might tip the balance. I felt as I was writing it that I might be missing something. Do you have a couple of minutes to double check the query engine [as it is in this branch](https://github.com/opensafely-core/databuilder/blob/benbc/generative-test-improvements/databuilder/query_engines/base_sql.py) to see if you can spot any opportunities for further simplification?

benbc commented on databuilder/query_model_transforms.py:
```diff
@@ -200,6 +201,27 @@ def get_dependers(node, nodes):
     return dependers
 
 
+def get_frame_operations(frame):
```
I don't think so. It's moved here from `query_model` as a deliberate part of this commit because it's only used here.

evansd replied:
Ah sorry, yes. It's moved around a bit and I got confused.

evansd commented:
@benbc I've taken another look at the code in that branch. And I'm remember spending a while thinking about this when first reviewing this. I don't feel like there are any massive simplifications lying around here (though I'm so used to a certain way of thinking about this code that I may well be missing something obvious).

My reasoning is that, one way or another, we need to walk a chain of operations and the only difference is that in one case the chain co-mingles two types of operation and in the other case the types are cleanly separated. But walking the co-mingled chain just isn't that much more algorithmically difficult (as I see it) than walking the separated one. Particularly if, for the sake of robustness, we decide we need some kind of assertion that the separated chain really _is_ separated.

Again, I might be missing something obvious, but it isn't obvious (yet) to me.

benbc commented:
@evansd Thank you. I've removed those changes. The commits were pretty much orthogonal to the rest of the PR and rebased out cleanly, so there is no need for you to re-review.
